# TypeScriptãƒ†ã‚¹ãƒˆãƒ«ãƒ¼ãƒ«

ã“ã®ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆã¯ã€zosqlãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã«ãŠã‘ã‚‹TypeScriptãƒ†ã‚¹ãƒˆã®æ¨™æº–ãƒ«ãƒ¼ãƒ«ã‚’å®šç¾©ã—ã¾ã™ã€‚

## ãƒ†ã‚¹ãƒˆãƒ•ãƒ¬ãƒ¼ãƒ ãƒ¯ãƒ¼ã‚¯

### ä½¿ç”¨ã™ã‚‹ãƒ†ã‚¹ãƒˆãƒ•ãƒ¬ãƒ¼ãƒ ãƒ¯ãƒ¼ã‚¯
- **Vitest** - ãƒ¢ãƒ€ãƒ³ã§é«˜é€Ÿãªãƒ†ã‚¹ãƒˆãƒ©ãƒ³ãƒŠãƒ¼
  - Viteãƒ™ãƒ¼ã‚¹ã®ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã¨ã®è¦ªå’Œæ€§ãŒé«˜ã„
  - Jest APIã¨ã®äº’æ›æ€§
  - ä¸¦åˆ—å®Ÿè¡Œã«ã‚ˆã‚‹é«˜é€ŸåŒ–
  - TypeScriptã®ã‚µãƒãƒ¼ãƒˆãŒçµ„ã¿è¾¼ã¿æ¸ˆã¿

## åŸºæœ¬æ–¹é‡

### ãƒ†ã‚¹ãƒˆä¿¡é ¼æ€§å„ªå…ˆã®æˆ¦ç•¥
- **å®Ÿè£…ã«æœ€ã‚‚è¿‘ã„å½¢ã§ãƒ†ã‚¹ãƒˆã™ã‚‹**
- **ãƒ¢ãƒƒã‚¯ã¯æœ¬å½“ã«å¿…è¦ãªå¤–éƒ¨ä¾å­˜ã®ã¿**
- **è»½é‡ãªå®Ÿè£…ã‚¯ãƒ©ã‚¹ã‚’ç©æ¥µæ´»ç”¨**

### ãƒ†ã‚¹ãƒˆæˆ¦ç•¥ã®ãƒ”ãƒ©ãƒŸãƒƒãƒ‰
```
        çµ±åˆãƒ†ã‚¹ãƒˆ (5%)
      â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
      â”‚  ãƒ–ãƒ©ã‚¦ã‚¶ãƒ†ã‚¹ãƒˆ  â”‚  â† æœ€å°é™
      â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
    â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
    â”‚   ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆãƒ†ã‚¹ãƒˆ  â”‚  â† UI bindingç¢ºèªã®ã¿ (15%)
    â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
  â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
  â”‚    å®Ÿè£…é‡è¦–å˜ä½“ãƒ†ã‚¹ãƒˆ       â”‚  â† ãƒ¡ã‚¤ãƒ³ãƒ•ã‚©ãƒ¼ã‚«ã‚¹ (80%)
  â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

### å“è³ªç›®æ¨™
- **å˜ä½“ãƒ†ã‚¹ãƒˆã‚«ãƒãƒ¬ãƒƒã‚¸: 90%ä»¥ä¸Š**
- **ãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯ã‚«ãƒãƒ¬ãƒƒã‚¸: 100%**
- **å®Ÿè¡Œæ™‚é–“: å˜ä½“ãƒ†ã‚¹ãƒˆå…¨ä½“ã§30ç§’ä»¥å†…**
- **ãƒ†ã‚¹ãƒˆã¨å®Ÿè£…ã®ä¹–é›¢æœ€å°åŒ–**

## ãƒ†ã‚¹ãƒˆãƒ•ã‚¡ã‚¤ãƒ«æ§‹æˆ

### ãƒ•ã‚¡ã‚¤ãƒ«å‘½åè¦å‰‡
```
src/
â”œâ”€â”€ core/
â”‚   â”œâ”€â”€ commands/
â”‚   â”‚   â”œâ”€â”€ execute-query-command.ts
â”‚   â”‚   â””â”€â”€ execute-query-command.test.ts     # åŒã˜ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒª
â”‚   â””â”€â”€ usecases/
â”‚       â”œâ”€â”€ sql-decomposer-usecase.ts
â”‚       â””â”€â”€ sql-decomposer-usecase.test.ts
â”œâ”€â”€ ui/
â”‚   â”œâ”€â”€ viewmodels/
â”‚   â”‚   â”œâ”€â”€ main-content-viewmodel.ts
â”‚   â”‚   â””â”€â”€ main-content-viewmodel.test.ts
â”‚   â””â”€â”€ components/
â”‚       â”œâ”€â”€ SqlEditor.tsx
â”‚       â””â”€â”€ SqlEditor.test.tsx               # æœ€å°é™ã®ãƒ†ã‚¹ãƒˆ
â””â”€â”€ test/
    â”œâ”€â”€ integration/                          # çµ±åˆãƒ†ã‚¹ãƒˆ
    â”‚   â””â”€â”€ sql-execution-flow.test.ts
    â””â”€â”€ helpers/                              # ãƒ†ã‚¹ãƒˆãƒ˜ãƒ«ãƒ‘ãƒ¼
        â”œâ”€â”€ mock-factories.ts
        â””â”€â”€ test-data.ts
```

### ãƒ•ã‚¡ã‚¤ãƒ«å‘½åãƒ‘ã‚¿ãƒ¼ãƒ³
- å˜ä½“ãƒ†ã‚¹ãƒˆ: `*.test.ts` ã¾ãŸã¯ `*.spec.ts`
- çµ±åˆãƒ†ã‚¹ãƒˆ: `*.integration.test.ts`
- E2Eãƒ†ã‚¹ãƒˆ: `*.e2e.test.ts`

## å˜ä½“ãƒ†ã‚¹ãƒˆãƒ«ãƒ¼ãƒ«

### å¿…é ˆå¯¾è±¡
ä»¥ä¸‹ã®ã‚³ãƒ¼ãƒ‰ã¯**å¿…ãš**å˜ä½“ãƒ†ã‚¹ãƒˆã‚’æ›¸ãï¼š

1. **Commandã‚¯ãƒ©ã‚¹** - ã™ã¹ã¦ã®ãƒ¡ã‚½ãƒƒãƒ‰
2. **UseCaseã‚¯ãƒ©ã‚¹** - ã™ã¹ã¦ã®ãƒ¡ã‚½ãƒƒãƒ‰
3. **ViewModelã‚¯ãƒ©ã‚¹** - ã™ã¹ã¦ã®ãƒ¡ã‚½ãƒƒãƒ‰ã¨ãƒ—ãƒ­ãƒ‘ãƒ†ã‚£
4. **Entityã‚¯ãƒ©ã‚¹** - ãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯ã‚’å«ã‚€ãƒ¡ã‚½ãƒƒãƒ‰
5. **ç´”ç²‹é–¢æ•°** - ã™ã¹ã¦ã®é–¢æ•°

### ãƒ†ã‚¹ãƒˆæ§‹é€ 
```typescript
import { describe, it, expect, beforeEach, vi } from 'vitest';

describe('ExecuteQueryCommand', () => {
  // ãƒ†ã‚¹ãƒˆå¯¾è±¡ã®æº–å‚™
  let command: ExecuteQueryCommand;
  let mockSqlExecutor: MockedObject<SqlExecutor>;
  let mockWorkspace: MockedObject<Workspace>;
  
  beforeEach(() => {
    // ãƒ¢ãƒƒã‚¯ã®åˆæœŸåŒ–ï¼ˆvi.mocked ã‚’ä½¿ç”¨ï¼‰
    mockSqlExecutor = createMockSqlExecutor();
    mockWorkspace = createMockWorkspace();
    
    command = new ExecuteQueryCommand({
      sql: 'SELECT * FROM users',
      sqlExecutor: mockSqlExecutor,
      workspace: mockWorkspace
    });
  });
  
  describe('canExecute', () => {
    it('should return true when SQL is not empty', () => {
      expect(command.canExecute()).toBe(true);
    });
    
    it('should return false when SQL is empty', () => {
      command = new ExecuteQueryCommand({
        sql: '',
        sqlExecutor: mockSqlExecutor,
        workspace: mockWorkspace
      });
      
      expect(command.canExecute()).toBe(false);
    });
  });
  
  describe('execute', () => {
    it('should execute SQL and return results', async () => {
      // Arrange
      const expectedResult = { rows: [{ id: 1, name: 'John' }] };
      mockSqlExecutor.execute.mockResolvedValue(expectedResult);
      
      // Act
      const result = await command.execute();
      
      // Assert
      expect(result).toEqual(expectedResult);
      expect(mockSqlExecutor.execute).toHaveBeenCalledWith('SELECT * FROM users');
    });
    
    it('should throw BusinessError when SQL is invalid', async () => {
      // Arrange
      mockSqlExecutor.execute.mockRejectedValue(new Error('Syntax error'));
      
      // Act & Assert
      await expect(command.execute()).rejects.toThrow(BusinessError);
    });
  });
});
```

### AAAï¼ˆArrange-Act-Assertï¼‰ãƒ‘ã‚¿ãƒ¼ãƒ³
```typescript
it('should format SQL with workspace formatter style', async () => {
  // Arrange - ãƒ†ã‚¹ãƒˆãƒ‡ãƒ¼ã‚¿ã®æº–å‚™
  const sql = 'SELECT id,name FROM users WHERE active=1';
  const formatterStyle = { indentSize: 2, uppercase: true };
  const command = new FormatSqlCommand(sql, formatterStyle);
  
  // Act - å®Ÿéš›ã®å®Ÿè¡Œ
  const result = await command.execute();
  
  // Assert - çµæœã®æ¤œè¨¼
  expect(result).toContain('SELECT');
  expect(result).toContain('FROM');
  expect(result).toContain('WHERE');
  expect(result.split('\n').length).toBeGreaterThan(1);
});
```

## ViewModelãƒ†ã‚¹ãƒˆãƒ«ãƒ¼ãƒ«

### å®Ÿè£…é‡è¦–ã®ViewModelãƒ†ã‚¹ãƒˆ
```typescript
describe('SqlEditorViewModel', () => {
  let viewModel: SqlEditorViewModel;
  let commandExecutor: CommandExecutor; // å®Ÿè£…ã‚’ä½¿ç”¨
  let workspace: Workspace;
  
  beforeEach(() => {
    // è»½é‡ãªå®Ÿè£…ã‚’ä½¿ç”¨
    commandExecutor = new CommandExecutor();
    workspace = createTestWorkspace(); // å®Ÿéš›ã®Workspaceã‚¨ãƒ³ãƒ†ã‚£ãƒ†ã‚£
    viewModel = new SqlEditorViewModel(commandExecutor, workspace);
  });
  
  describe('sql property', () => {
    it('should notify change when sql is updated', () => {
      const changeListener = vi.fn();
      viewModel.subscribe('sql', changeListener);
      
      viewModel.sql = 'SELECT * FROM users';
      
      expect(changeListener).toHaveBeenCalledWith('sql');
      expect(viewModel.sql).toBe('SELECT * FROM users');
    });
  });
  
  describe('formatQuery', () => {
    it('should format SQL using real formatter', async () => {
      // Arrange
      viewModel.sql = 'SELECT id,name FROM users WHERE active=1';
      
      // Act - å®Ÿéš›ã®ãƒ•ã‚©ãƒ¼ãƒãƒƒãƒˆå‡¦ç†ã‚’å®Ÿè¡Œ
      await viewModel.formatQuery();
      
      // Assert - å®Ÿéš›ã«ãƒ•ã‚©ãƒ¼ãƒãƒƒãƒˆã•ã‚Œã¦ã„ã‚‹ã‹ãƒã‚§ãƒƒã‚¯
      expect(viewModel.sql).toContain('\n'); // æ”¹è¡ŒãŒå…¥ã£ã¦ã„ã‚‹
      expect(viewModel.sql).toContain('SELECT');
      expect(viewModel.sql).toContain('FROM');
      expect(viewModel.sql).toContain('WHERE');
    });
  });
  
  describe('executeQuery', () => {
    it('should manage execution state correctly', async () => {
      // Arrange
      viewModel.sql = 'SELECT 1 as test';
      
      // Act
      const executePromise = viewModel.executeQuery();
      
      // Assert - å®Ÿè¡Œä¸­ã®çŠ¶æ…‹
      expect(viewModel.isExecuting).toBe(true);
      expect(viewModel.canExecute).toBe(false);
      
      await executePromise;
      
      // Assert - å®Ÿè¡Œå¾Œã®çŠ¶æ…‹
      expect(viewModel.isExecuting).toBe(false);
      expect(viewModel.canExecute).toBe(true);
    });
  });
});

// ãƒ†ã‚¹ãƒˆãƒ˜ãƒ«ãƒ‘ãƒ¼: å®Ÿéš›ã®ã‚¨ãƒ³ãƒ†ã‚£ãƒ†ã‚£ã‚’ä½œæˆ
function createTestWorkspace(): Workspace {
  return new Workspace({
    id: 'test-workspace',
    name: 'Test Workspace',
    formatterStyle: {
      indentSize: 2,
      uppercase: true,
      linesBetweenQueries: 1
    }
  });
}
```

## ãƒ†ã‚¹ãƒˆãƒ€ãƒ–ãƒ«æˆ¦ç•¥

### å„ªå…ˆé †ä½ï¼šå®Ÿè£… > è»½é‡å®Ÿè£… > ãƒ¢ãƒƒã‚¯

```typescript
// ğŸ¥‡ æœ€å„ªå…ˆ: å®Ÿè£…ã‚’ãã®ã¾ã¾ä½¿ç”¨
describe('SqlFormatterCommand', () => {
  it('should format SQL using real rawsql-ts', async () => {
    const command = new FormatSqlCommand(
      'SELECT * FROM users',
      { indentSize: 2, uppercase: true }
    );
    
    // å®Ÿéš›ã®rawsql-tsã‚’ä½¿ç”¨
    const result = await command.execute();
    expect(result).toContain('SELECT');
  });
});

// ğŸ¥ˆ æ¬¡å„ªå…ˆ: è»½é‡ãªãƒ†ã‚¹ãƒˆå®Ÿè£…
class InMemoryWorkspaceRepository implements WorkspaceRepositoryPort {
  private workspaces = new Map<string, Workspace>();
  
  async save(workspace: Workspace): Promise<void> {
    this.workspaces.set(workspace.id, workspace);
  }
  
  async findById(id: string): Promise<Workspace | null> {
    return this.workspaces.get(id) || null;
  }
}

// ğŸ¥‰ æœ€å¾Œã®æ‰‹æ®µ: å¤–éƒ¨ä¾å­˜ã®ã¿ãƒ¢ãƒƒã‚¯
describe('FileSystemService', () => {
  let mockFileSystem: MockedObject<FileSystem>;
  
  beforeEach(() => {
    // ãƒ•ã‚¡ã‚¤ãƒ«ã‚·ã‚¹ãƒ†ãƒ ã¯å®Ÿéš›ã«ã‚¢ã‚¯ã‚»ã‚¹ã§ããªã„ãŸã‚ãƒ¢ãƒƒã‚¯
    mockFileSystem = {
      readFile: vi.fn(),
      writeFile: vi.fn(),
      exists: vi.fn()
    };
  });
});
```

### ãƒ¢ãƒƒã‚¯ä½¿ç”¨ã®åˆ¤æ–­åŸºæº–

#### âœ… ãƒ¢ãƒƒã‚¯ã—ã¦ã‚ˆã„ã‚‚ã®
- **ãƒ–ãƒ©ã‚¦ã‚¶API**: localStorage, IndexedDB, FileSystem
- **é‡ã„å¤–éƒ¨ãƒ©ã‚¤ãƒ–ãƒ©ãƒª**: PGliteã®å®Ÿéš›ã®SQLå®Ÿè¡Œ
- **éåŒæœŸI/O**: ãƒãƒƒãƒˆãƒ¯ãƒ¼ã‚¯é€šä¿¡ã€ãƒ•ã‚¡ã‚¤ãƒ«èª­ã¿æ›¸ã
- **æ™‚é–“ä¾å­˜**: Date.now(), setTimeout
- **ãƒ©ãƒ³ãƒ€ãƒ æ€§**: Math.random()

#### âŒ ãƒ¢ãƒƒã‚¯ã—ã¦ã¯ã„ã‘ãªã„ã‚‚ã®
- **è‡ªä½œã®ãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯**: Entityã‚¯ãƒ©ã‚¹ã€UseCaseã‚¯ãƒ©ã‚¹
- **è»½é‡ãªãƒ©ã‚¤ãƒ–ãƒ©ãƒª**: rawsql-tsã®ãƒ‘ãƒ¼ã‚µãƒ¼éƒ¨åˆ†
- **ç´”ç²‹é–¢æ•°**: è¨ˆç®—å‡¦ç†ã€ãƒ•ã‚©ãƒ¼ãƒãƒƒãƒˆå‡¦ç†
- **ãƒ‡ãƒ¼ã‚¿æ§‹é€ **: é…åˆ—æ“ä½œã€ã‚ªãƒ–ã‚¸ã‚§ã‚¯ãƒˆå¤‰æ›

### å®Ÿè£…é‡è¦–ã®ãƒ†ã‚¹ãƒˆä¾‹
```typescript
// âœ… æ¨å¥¨: å®Ÿéš›ã®rawsql-tsã‚’ä½¿ç”¨
describe('SqlDecomposerUseCase', () => {
  let usecase: SqlDecomposerUseCase;
  let repository: InMemoryWorkspaceRepository;
  
  beforeEach(() => {
    repository = new InMemoryWorkspaceRepository();
    usecase = new SqlDecomposerUseCase(repository);
  });
  
  it('should decompose CTE using real parser', async () => {
    const sql = `
      WITH user_stats AS (
        SELECT user_id, COUNT(*) FROM orders GROUP BY user_id
      )
      SELECT * FROM user_stats
    `;
    
    // å®Ÿéš›ã®ãƒ‘ãƒ¼ã‚µãƒ¼ã§ãƒ†ã‚¹ãƒˆ
    const result = await usecase.decompose(sql);
    
    expect(result.ctes).toHaveLength(1);
    expect(result.ctes[0].name).toBe('user_stats');
    expect(result.ctes[0].sql).toContain('SELECT user_id, COUNT(*)');
  });
});

// âŒ é¿ã‘ã‚‹: ä¸è¦ãªãƒ¢ãƒƒã‚¯
describe('SqlDecomposerUseCase', () => {
  let mockParser: MockedObject<SqlParser>;
  
  beforeEach(() => {
    mockParser = {
      parse: vi.fn().mockReturnValue(fakeParsedResult)
    };
  });
  
  // ã“ã®ãƒ†ã‚¹ãƒˆã¯å®Ÿéš›ã®ãƒ‘ãƒ¼ã‚µãƒ¼ã®å‹•ä½œã‚’æ¤œè¨¼ã—ã¦ã„ãªã„
});
```

## UIã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆãƒ†ã‚¹ãƒˆ

### æœ€å°é™ã®çµ±åˆãƒ†ã‚¹ãƒˆ
```typescript
// ui/components/SqlEditor.test.tsx
import { render, screen, fireEvent } from '@testing-library/react';
import { vi } from 'vitest';

describe('SqlEditor Component', () => {
  let mockViewModel: MockedObject<SqlEditorViewModel>;
  
  beforeEach(() => {
    mockViewModel = {
      sql: '',
      canExecute: false,
      isExecuting: false,
      executeQuery: vi.fn(),
      formatQuery: vi.fn(),
      subscribe: vi.fn(),
      unsubscribe: vi.fn()
    };
  });
  
  it('should render monaco editor with viewmodel sql', () => {
    mockViewModel.sql = 'SELECT * FROM users';
    
    render(<SqlEditor viewModel={mockViewModel} />);
    
    expect(screen.getByDisplayValue('SELECT * FROM users')).toBeInTheDocument();
  });
  
  it('should call viewModel.executeQuery when execute button is clicked', () => {
    mockViewModel.canExecute = true;
    
    render(<SqlEditor viewModel={mockViewModel} />);
    fireEvent.click(screen.getByText('Execute'));
    
    expect(mockViewModel.executeQuery).toHaveBeenCalled();
  });
  
  // âŒ é¿ã‘ã‚‹: è©³ç´°ãªUIæ“ä½œãƒ†ã‚¹ãƒˆ
  // it('should show loading spinner during execution', () => { ... });
});
```

## ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ãƒ†ã‚¹ãƒˆ

### å®Ÿè¡Œæ™‚é–“ã®æ¸¬å®š
```typescript
describe('Performance Tests', () => {
  it('should parse large SQL within time limit', async () => {
    const largeSql = generateLargeSql(1000); // 1000è¡Œã®SQL
    const startTime = performance.now();
    
    const command = new ParseSqlCommand(largeSql);
    await command.execute();
    
    const endTime = performance.now();
    const executionTime = endTime - startTime;
    
    expect(executionTime).toBeLessThan(100); // 100msä»¥å†…
  });
});
```

## ãƒ†ã‚¹ãƒˆãƒ‡ãƒ¼ã‚¿ç®¡ç†

### ãƒ†ã‚¹ãƒˆãƒ•ã‚£ã‚¯ã‚¹ãƒãƒ£
```typescript
// test/helpers/test-data.ts
export const TEST_SQL_QUERIES = {
  simple: 'SELECT id, name FROM users',
  withWhere: 'SELECT id, name FROM users WHERE active = 1',
  withCte: `
    WITH active_users AS (
      SELECT id, name FROM users WHERE active = 1
    )
    SELECT * FROM active_users
  `,
  complex: `
    WITH user_stats AS (
      SELECT user_id, COUNT(*) as order_count
      FROM orders 
      GROUP BY user_id
    ),
    high_value_users AS (
      SELECT user_id FROM user_stats WHERE order_count > 10
    )
    SELECT u.name, us.order_count
    FROM users u
    JOIN user_stats us ON u.id = us.user_id
    WHERE u.id IN (SELECT user_id FROM high_value_users)
  `
};

export const TEST_WORKSPACES = {
  default: {
    id: 'default',
    name: 'Default Workspace',
    formatterStyle: DEFAULT_FORMATTER_STYLE
  },
  custom: {
    id: 'custom',
    name: 'Custom Workspace',
    formatterStyle: {
      indentSize: 4,
      uppercase: false,
      linesBetweenQueries: 2
    }
  }
};
```

## ãƒ†ã‚¹ãƒˆå®Ÿè¡Œãƒ«ãƒ¼ãƒ«

### å®Ÿè¡Œã‚³ãƒãƒ³ãƒ‰
```bash
# å…¨ãƒ†ã‚¹ãƒˆå®Ÿè¡Œ
npm run test:run

# ã‚¦ã‚©ãƒƒãƒãƒ¢ãƒ¼ãƒ‰
npm run test

# UIãƒ¢ãƒ¼ãƒ‰ã§ãƒ†ã‚¹ãƒˆå®Ÿè¡Œ
npm run test:ui

# ã‚«ãƒãƒ¬ãƒƒã‚¸ä»˜ãå®Ÿè¡Œ
npm run test:coverage

# ç‰¹å®šã®ãƒ†ã‚¹ãƒˆãƒ•ã‚¡ã‚¤ãƒ«
npm run test sql-editor

# ç‰¹å®šã®ãƒ†ã‚¹ãƒˆãƒ‘ã‚¿ãƒ¼ãƒ³
npm run test -- -t "should execute"
```

### CI/CDè¦ä»¶
```yaml
# .github/workflows/test.yml
- name: Run Unit Tests
  run: |
    npm run test:run
    npm run test:coverage
  env:
    NODE_ENV: test
    
- name: Check Coverage Thresholds
  run: |
    # ã‚«ãƒãƒ¬ãƒƒã‚¸ãŒ90%æœªæº€ã®å ´åˆã¯å¤±æ•—
    npm run test:coverage:check
```

## ç¦æ­¢äº‹é …

### ã‚„ã£ã¦ã¯ã„ã‘ãªã„ãƒ†ã‚¹ãƒˆ
```typescript
// âŒ å®Ÿè£…è©³ç´°ã‚’ãƒ†ã‚¹ãƒˆã—ãªã„
it('should call internal method', () => {
  const spy = vi.spyOn(service, '_internalMethod');
  service.execute();
  expect(spy).toHaveBeenCalled();
});

// âŒ å¤–éƒ¨ã‚µãƒ¼ãƒ“ã‚¹ã«å®Ÿéš›ã«ã‚¢ã‚¯ã‚»ã‚¹
it('should save to database', async () => {
  await service.save(data);
  const saved = await database.find(data.id);
  expect(saved).toEqual(data);
});

// âŒ æ™‚é–“ã«ä¾å­˜ã™ã‚‹ãƒ†ã‚¹ãƒˆ
it('should return current timestamp', () => {
  const result = service.getCurrentTime();
  expect(result).toBe(Date.now()); // ä¸å®‰å®š
});

// âŒ é †åºã«ä¾å­˜ã™ã‚‹ãƒ†ã‚¹ãƒˆ
describe('Sequential Tests', () => {
  let sharedState: any;
  
  it('should set state', () => {
    sharedState = { value: 1 };
  });
  
  it('should use previous state', () => {
    expect(sharedState.value).toBe(1); // å‰ã®ãƒ†ã‚¹ãƒˆã«ä¾å­˜
  });
});
```

### æ¨å¥¨ã•ã‚Œã‚‹ä»£æ›¿æ¡ˆ
```typescript
// âœ… æŒ¯ã‚‹èˆã„ã‚’ãƒ†ã‚¹ãƒˆ
it('should return formatted SQL', async () => {
  const result = await service.format(inputSql);
  expect(result).toContain('SELECT');
  expect(result).toContain('FROM');
});

// âœ… ãƒ¢ãƒƒã‚¯ã‚’ä½¿ç”¨
it('should call save with correct data', async () => {
  const mockDatabase = vi.fn();
  const service = new Service(mockDatabase);
  
  await service.save(data);
  expect(mockDatabase).toHaveBeenCalledWith(data);
});

// âœ… æ™‚é–“ã‚’ãƒ¢ãƒƒã‚¯
it('should return current timestamp', () => {
  const fixedTime = new Date('2023-01-01').getTime();
  vi.spyOn(Date, 'now').mockReturnValue(fixedTime);
  
  const result = service.getCurrentTime();
  expect(result).toBe(fixedTime);
});

// âœ… ç‹¬ç«‹ã—ãŸãƒ†ã‚¹ãƒˆ
describe('Independent Tests', () => {
  let service: Service;
  
  beforeEach(() => {
    service = new Service();
  });
  
  it('should handle empty input', () => {
    const result = service.process('');
    expect(result).toEqual(EMPTY_RESULT);
  });
  
  it('should handle valid input', () => {
    const result = service.process('valid input');
    expect(result).toEqual(EXPECTED_RESULT);
  });
});
```

## ãƒ‡ãƒãƒƒã‚°ã¨ãƒˆãƒ©ãƒ–ãƒ«ã‚·ãƒ¥ãƒ¼ãƒ†ã‚£ãƒ³ã‚°

### ãƒ†ã‚¹ãƒˆå¤±æ•—æ™‚ã®èª¿æŸ»æ‰‹é †
```typescript
// ãƒ‡ãƒãƒƒã‚°ç”¨ã®ãƒ­ã‚°å‡ºåŠ›
it('should process complex query', async () => {
  const command = new ProcessQueryCommand(complexSql);
  
  // ãƒ‡ãƒãƒƒã‚°æƒ…å ±ã‚’å‡ºåŠ›
  console.log('Input SQL:', complexSql);
  console.log('Command state:', command.getState());
  
  const result = await command.execute();
  
  console.log('Result:', result);
  expect(result.success).toBe(true);
});

// ãƒ¢ãƒƒã‚¯ã®å‘¼ã³å‡ºã—å±¥æ­´ã‚’ç¢ºèª
it('should call dependencies in correct order', async () => {
  await service.execute();
  
  expect(mockParser.parse).toHaveBeenCalledBefore(mockValidator.validate);
  expect(mockValidator.validate).toHaveBeenCalledBefore(mockExecutor.execute);
});
```

### éåŒæœŸãƒ†ã‚¹ãƒˆã®ãƒ‡ãƒãƒƒã‚°
```typescript
it('should handle async operations', async () => {
  const promise = service.asyncOperation();
  
  // Promiseã®çŠ¶æ…‹ã‚’ç¢ºèª
  expect(promise).toBeInstanceOf(Promise);
  
  const result = await promise;
  expect(result).toBeDefined();
});

// ã‚¿ã‚¤ãƒ ã‚¢ã‚¦ãƒˆã®è¨­å®š
it('should complete within timeout', async () => {
  const result = await service.longRunningOperation();
  expect(result).toBeDefined();
}, 10000); // 10ç§’ã‚¿ã‚¤ãƒ ã‚¢ã‚¦ãƒˆ
```

## å“è³ªãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ

### ãƒ—ãƒ«ãƒªã‚¯ã‚¨ã‚¹ãƒˆå‰ã®ç¢ºèªäº‹é …
- [ ] æ–°è¦è¿½åŠ ã—ãŸãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯ã«å˜ä½“ãƒ†ã‚¹ãƒˆã‚’è¿½åŠ 
- [ ] ã‚«ãƒãƒ¬ãƒƒã‚¸ãŒ90%ä»¥ä¸Šã‚’ç¶­æŒ
- [ ] ã™ã¹ã¦ã®ãƒ†ã‚¹ãƒˆãŒç‹¬ç«‹ã—ã¦å®Ÿè¡Œå¯èƒ½
- [ ] ãƒ¢ãƒƒã‚¯ãŒé©åˆ‡ã«ä½¿ç”¨ã•ã‚Œã¦ã„ã‚‹
- [ ] ãƒ†ã‚¹ãƒˆåãŒæŒ¯ã‚‹èˆã„ã‚’è¡¨ç¾ã—ã¦ã„ã‚‹
- [ ] AAAï¼ˆArrange-Act-Assertï¼‰ãƒ‘ã‚¿ãƒ¼ãƒ³ã«å¾“ã£ã¦ã„ã‚‹
- [ ] å®Ÿè£…è©³ç´°ã§ã¯ãªãæŒ¯ã‚‹èˆã„ã‚’ãƒ†ã‚¹ãƒˆã—ã¦ã„ã‚‹
- [ ] éåŒæœŸå‡¦ç†ãŒé©åˆ‡ã«ãƒ†ã‚¹ãƒˆã•ã‚Œã¦ã„ã‚‹

### å®šæœŸçš„ãªãƒ¡ãƒ³ãƒ†ãƒŠãƒ³ã‚¹
- **æœˆæ¬¡**: ãƒ†ã‚¹ãƒˆå®Ÿè¡Œæ™‚é–“ã‚’è¨ˆæ¸¬ã—ã€30ç§’ä»¥å†…ã‚’ç¶­æŒ
- **å››åŠæœŸ**: ä¸è¦ã«ãªã£ãŸãƒ†ã‚¹ãƒˆã®å‰Šé™¤
- **åŠå¹´**: ãƒ†ã‚¹ãƒˆãƒ˜ãƒ«ãƒ‘ãƒ¼ã¨ãƒ¢ãƒƒã‚¯ãƒ•ã‚¡ã‚¯ãƒˆãƒªã®è¦‹ç›´ã—

## æ›´æ–°å±¥æ­´
- 2025-01-27: åˆç‰ˆä½œæˆ
- 2025-01-27: Vitestã®ä½¿ç”¨ã‚’æ˜è¨˜ã€Jesté–¢æ•°ã‚’Vitesté–¢æ•°ã«æ›´æ–°